084-open-feature-integration.md

```diff
@@ -0,0 +1,477 @@
+
+On the other side for the more complex users with need of flagging system, we would support of [FlagD](https://github.com/open-feature/flagd),
+which is pretty known because [OpenFeature is CNCF Incubating project](https://www.cncf.io/projects/openfeature/).
+Common characteristics of this provider are:
+1. New method proposed for more complex users, which needs better management over multiple features (of Strimzi)
+2. Integrates with FlagD for dynamic feature flagging
+3. Allows changes without the need for rolling updates.
+
+### Implementation Steps
+
+Currently, there are a few classes where we use `FeatureGates` related stuff (i.e., FeatureGates abstraction).
+Those classes are:
+1. ClusterOperatorConfig 
+2. EntityTopicOperator 
+3. EntityUserOperator 
+4. AbstractConnectOperator
+5. KafkaReconciler
+6. ZooKeeperReconciler
+
+In each of these classes, we essentially do the following:
+
+```java
+result.featureGatesEnvVarValue = config.featureGates().toEnvironmentVariable()
+```
+
+Here, we parse the currently configured `FeatureGates` from the `STRIMZI_FEATURE_GATES` environment variable and assign it to specific classes (e.g., the `EntityTopicOperator` class).
+Furthermore, if we dive deeper into the implementation of the `UserOperator` class, we retrieve the feature gates as follows:
+```java
+/**
+ * @return Feature gates configuration
+ */
+public FeatureGates featureGates() {
+    return get(FEATURE_GATES);
+}
+```
+These are fetched directly from `ConfigParameter`, which encapsulates a map implementation with additional utility methods (e.g., a parser to convert the string representation into a specified type).
+In this case, the `FeatureGates` type is defined as:
+```java
+ /**
+ * Configuration string with feature gates settings
+ */
+public static final ConfigParameter<FeatureGates> FEATURE_GATES = new ConfigParameter<>("STRIMZI_FEATURE_GATES", parseFeatureGates(), "", CONFIG_VALUES);
+```
+Here, the `parseFeatureGates()` method is defined, which calls the constructor of the `FeatureGates` class.
+
+This is the basic flow of how it currently works, and my proposal is to modify the `FeatureGates` class to use `EnvVarProvider`, which would cover all aspects of the current implementation. 
+For this, we would need to add a few dependencies (e.g., `dev.openfeature.sdk` and `dev.openfeature.contrib.providers.env-var`).
+Additionally, we can easily extend support by adding other providers like `FlagDProvider` for users who want to use a feature flagging system (i.e., FlagD with its `OpenFeature` Operator). 
+
+The overall setup would look like this:
+
+```java
+// ...
+// imports omitted for brevity
+// ...
+
+/**
+ * Class for handling the configuration of feature gates
+ */
+public class FeatureGates {
+
+    private static final String CONTINUE_ON_MANUAL_RU_FAILURE = "ContinueReconciliationOnManualRollingUpdateFailure";
+    private static final String FLAGD_ENABLED_ENV_VAR = "FLAGD_ENABLED"; // Environment variable to toggle FlagD
```

Comment:
Again, why does this not configure the _provider_ instead of enabling a particular provider?

Reply:
Sure, this should be the provider and not just focus on FlagD :). Thanks